Spring-2019/system_design_misc_coding/SOLUTIONS_2019Spring.md

```diff
@@ -86,20 +88,50 @@ TODO :bug:
 
 TODO :bug:
 
-## 3. PROBLEM 3 TODO :bug:
+### 3. TinyURL
 
-Source: TODO :bug:
+Source: CTCI Chapter 9
 
-#### Naive/Simple Solution 
+#### Taking a First Crack at the Problem
 
-TODO :bug:
+Here's a rough sketch of ideas, potential system components, and 
+design considerations:
 
-#### Optimal Solution
+![system design sketch](./tinyurl/tiny-url-draft.jpg)
 
-TODO :bug:
+#### A Short Writeup
 
-#### Testing The Solutions OR Driver For Solution
+TinyURL would have 2 primary classes of customers:
 
-TODO :bug:
+- Users who generate and manage links
+- Individuals who receive the distributed links and visit the link (and 
+  get redirected to the target destination)
+
+To serve the first class of customers effectively, a web-based UI with 
+registration and account persistence makes the most sense. Users can log 
+in, create links, and update/delete links as desired.
```

Comment:
> a web-based UI with 
registration and account persistence makes the most sense.

Although most of the URL shorteners I researched (tinyURL, Bit.ly, etc) do not require account registration to shorten URL links. How would you handle this use case? Would it impact the current design significantly?

Reply:
I think I have addressed this, but if you see any discrepancies, lmk. see 7b4bca942b4bad6fca60f49e9660d468a8cd748a for new changes